main/compiling/compilers/Compiler.py

```diff
@@ -97,48 +103,129 @@ def __init__(
-                initial state.
-            initial_stabilizers: If initial_states is not provided,
-                this can be used to determine the initial states of the data
-                qubits.
-            final_measurements: The measurements to perform at the end of the
-                circuit. If not provided, will use final_stabilizers to
-                determine the measurements.
-            final_stabilizers: If final_measurements is not provided, this can
-                be used to determine the measurements to perform at the end of
-                the circuit.
-            logical_observables: The logical observables to construnct from
-                the final measurements.
-
-                                
-        Raises:
-            ValueError: If neither initial_states nor initial_stabilizers are
-                provided.
-            ValueError: If both final_measurements and final_stabilizers are
-                provided.
-            ValueError: If the number of layers is not a positive integer.
+        self.check_validity_of_inputs(
+            code, initial_states, initial_stabilizers, final_measurements, final_stabilizers, observables
+        )
+        initial_detector_schedules, tick, circuit = self.compile_initialisation(
+            code, initial_states, initial_stabilizers)
 
-        Returns:
-            An instance of the Circuit class.
+        initialization_layers = len(initial_detector_schedules)
+        # initial_layers is the number of layers in which 'lid-only'
+        # detectors exist.
+        if initialization_layers * code.schedule_length > total_rounds:
+            raise ValueError(
+                f"The number of layers required to set up the code is "
+                f"greater than the number of layers to compile!"
+                f"Requested that {total_rounds} round(s) are compiled, but code "
+                f"seems to take {initialization_layers * code.schedule_length} round(s) to set up.")
+
+        # Compile these initial layers.
+        for layer, detector_schedule in enumerate(initial_detector_schedules):
+            tick = self.compile_layer(
+                layer, detector_schedule, observables, tick, circuit, code
+            )
+
+        # Compile the remaining layers.
+        round = code.schedule_length * initialization_layers
+        while round < total_rounds:
+            tick = self.compile_round(
+                round,
+                round % code.schedule_length,
+
+                code.detector_schedule,
+                observables,
+                tick,
+                circuit,
+                code,
+            )
+            round += 1
+
+        # Finish with data qubit measurements, and use these to reconstruct
+        # some detectors.
+        if isinstance(code, TicTacToeCode) or isinstance(code, GaugeTicTacToeCode):
+            pauli_letter_observable = observables[0].at_round(round-1)[
```

Comment:
Add comment explaining that it's okay to just use observables[0] (i.e. to assume there's only one observable) because it's within a hack and we've accepted we're spaghetti-ing the code a bit